src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go

```diff
@@ -3,22 +3,150 @@ package writers
-		"eventType": eventType,
-		"userID":    userID,
-	}).Info("Doing stuff...")
-	return util.MessageResponse(404, "Not implemented yet")
+	var r map[string]interface{} // must be a JSON object
+	resErr = httputil.UnmarshalJSONRequest(req, &r)
+	if resErr != nil {
+		return *resErr
+	}
+
+	// create the new event and set all the fields we can
+	builder := gomatrixserverlib.EventBuilder{
+		Sender:   userID,
+		RoomID:   roomID,
+		Type:     eventType,
+		StateKey: nil,
+	}
+	builder.SetContent(r)
+
+	// work out what will be required in order to send this event
+	requiredStateEvents, err := stateNeeded(&builder)
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	// Ask the roomserver for information about this room
+	queryReq := api.QueryLatestEventsAndStateRequest{
+		RoomID:       roomID,
+		StateToFetch: requiredStateEvents,
+	}
+	var queryRes api.QueryLatestEventsAndStateResponse
+	if queryErr := queryAPI.QueryLatestEventsAndState(&queryReq, &queryRes); queryErr != nil {
+		return httputil.LogThenError(req, queryErr)
+	}
+	if !queryRes.RoomExists {
+		return util.JSONResponse{
+			Code: 404,
+			JSON: jsonerror.NotFound("Room does not exist"),
+		}
+	}
+
+	// set the fields we previously couldn't do and build the event
+	builder.PrevEvents = queryRes.LatestEvents // the current events will be the prev events of the new event
+	var refs []gomatrixserverlib.EventReference
+	for _, e := range queryRes.StateEvents {
+		refs = append(refs, e.EventReference())
+	}
+	builder.AuthEvents = refs
+	eventID := fmt.Sprintf("$%s:%s", util.RandomString(16), cfg.ServerName)
+	e, err := builder.Build(eventID, time.Now(), cfg.ServerName, cfg.KeyID, cfg.PrivateKey)
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	// check to see if this user can perform this operation
+	stateEvents := make([]*gomatrixserverlib.Event, len(queryRes.StateEvents))
+	for i := range queryRes.StateEvents {
+		stateEvents[i] = &queryRes.StateEvents[i]
+	}
+	provider := gomatrixserverlib.NewAuthEvents(stateEvents)
+	if err = gomatrixserverlib.Allowed(e, &provider); err != nil {
+		return util.JSONResponse{
+			Code: 403,
+			JSON: jsonerror.Forbidden(err.Error()), // TODO: Is this error string comprehensible to the client?
+		}
+	}
+
+	// pass the new event to the roomserver
+	if err := sendToRoomserver(e, producer, cfg.ClientAPIOutputTopic); err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	return util.JSONResponse{
+		Code: 200,
+		JSON: sendMessageResponse{e.EventID()},
+	}
+}
+
+func sendToRoomserver(e gomatrixserverlib.Event, producer sarama.SyncProducer, topic string) error {
+	var authEventIDs []string
+	for _, ref := range e.AuthEvents() {
+		authEventIDs = append(authEventIDs, ref.EventID)
+	}
+	ire := api.InputRoomEvent{
+		Kind:         api.KindNew,
+		Event:        e.JSON(),
+		AuthEventIDs: authEventIDs,
+	}
+
+	value, err := json.Marshal(ire)
+	if err != nil {
+		return err
+	}
+	var m sarama.ProducerMessage
```

Comment:
Instead of passing in a `sarama.Producer`, I'm happy to pass in an interface which will send `InputRoomEvents` to the roomserver. I'd like to do that in another PR though.